File: educational-core/src/com/jetbrains/edu/learning/actions/PrevPlaceholderAction.java
package com.jetbrains.edu.learning.actions;

import com.jetbrains.edu.learning.courseFormat.AnswerPlaceholder;
import com.jetbrains.edu.learning.courseFormat.TaskFile;
import kotlin.collections.CollectionsKt;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

public class PrevPlaceholderAction extends PlaceholderNavigationAction {
  @NonNls
  public static final String ACTION_ID = "Educational.PrevPlaceholder";

  @Nullable
  @Override
  protected AnswerPlaceholder getTargetPlaceholder(@NotNull final TaskFile taskFile, int offset) {
    final AnswerPlaceholder selectedAnswerPlaceholder = taskFile.getAnswerPlaceholder(offset);
    final List<AnswerPlaceholder> placeholders = taskFile.getAnswerPlaceholders();
    int endIndex = selectedAnswerPlaceholder != null ? selectedAnswerPlaceholder.getIndex() : placeholders.size();
    if (!indexIsValid(endIndex - 1, placeholders)) return null;

    for (AnswerPlaceholder placeholder : CollectionsKt.asReversed(placeholders.subList(0, endIndex))) {
      if (placeholder.getOffset() < offset && placeholder.isVisible()) {
        return placeholder;
      }
    }

    return null;
  }
}

File: intellij-plugin/educational-core/src/com/jetbrains/edu/learning/actions/PrevPlaceholderAction.kt
package com.jetbrains.edu.learning.actions

import com.intellij.openapi.actionSystem.ActionUpdateThread
import com.jetbrains.edu.learning.courseFormat.AnswerPlaceholder
import com.jetbrains.edu.learning.courseFormat.TaskFile
import org.jetbrains.annotations.NonNls

class PrevPlaceholderAction : PlaceholderNavigationAction() {
  override fun getTargetPlaceholder(taskFile: TaskFile, offset: Int): AnswerPlaceholder? {
    val selectedAnswerPlaceholder = taskFile.getAnswerPlaceholder(offset)
    val placeholders = taskFile.answerPlaceholders
    val endIndex = selectedAnswerPlaceholder?.index ?: placeholders.size
    if (!indexIsValid(endIndex - 1, placeholders)) return null
    for (placeholder in placeholders.subList(0, endIndex).asReversed()) {
      if (placeholder.offset < offset && placeholder.isCurrentlyVisible) {
        return placeholder
      }
    }
    return null
  }

  override fun getActionUpdateThread(): ActionUpdateThread = ActionUpdateThread.BGT

  companion object {
    const val ACTION_ID: @NonNls String = "Educational.PrevPlaceholder"
  }
}
